File: src/languages/telugu/telugu.ts
/**
 * Native Telugu (te) text phonemizer — canonical IPA, espeak-independent. Telugu is a Dravidian Brahmic abugida
 * read by the generic engine (core/abugida.ts); unlike Hindi there is NO inherent-vowel deletion (every akshara
 * is pronounced — inherent /a/). telugu.ts adds only the light post-processing: geminate → length, and the
 * word-final anusvara ం realized as [m] (అంకురం → aŋkuɾam). First-syllable stress (weak; the backbone folds it).
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { LATIN_RUN } from "../../core/hostWord.ts";
import { makeAbugidaG2P } from "../../core/abugida.ts";
import { loadSharedPhonology } from "../../core/phonology.ts";
import { MANIFEST } from "./manifest.ts";
import { numberToWords } from "./numbers.ts";
import { normalizeTelugu } from "./normalize.ts";

const DEF = MANIFEST;
const CLAUSE_MARK = DEF.clausePunctuation;
const TELUGU_WORD = "ఀ-౯ౠ-ౣ";
const TELUGU_DIGITS: Record<string, string> = {
    "౦": "0", "౧": "1", "౨": "2", "౩": "3", "౪": "4",
    "౫": "5", "౬": "6", "౭": "7", "౮": "8", "౯": "9",
};
const DIGIT_CLASS = "0-9" + Object.keys(TELUGU_DIGITS).join("");
const VOWEL = "aeiouɾ"; // nucleus starts (for the geminate/stress scan; ɾ only in the ɾu vocalic-r nucleus)

let G2P: ((w: string) => string) | undefined;
function g2p(w: string): string {
    return (G2P ??= makeAbugidaG2P(DEF, loadSharedPhonology()))(w);
}

// Geminate consonant (doubled base, possibly aspirated) → single + length ː.
const GEMINATE =
    /(t͡ʃʰ|d͡ʒʱ|t͡ʃ|d͡ʒ|t͡s|d͡z|t̪ʰ|d̪ʱ|ʈʰ|ɖʱ|ɡʱ|kʰ|t̪|d̪|[kɡpbmnlʃʂsʈɖɳɭɲŋjɦʋɾr])\1(?!͡)/gu;

/** One Telugu word → canonical IPA. */
export function phonemizeWord(word: string): string {
    // Word-final anusvara ం → [m] (అంకురం→aŋkuɾam); medial ం is a homorganic nasal, handled by the engine.
    const norm = word.normalize("NFC").replace(/ం$/u, "మ్");
    let x = g2p(norm);
    x = x.replace(GEMINATE, "$1ː").replace(/ː([ʰʱ])/gu, "$1ː");
    x = x.replace(/ɭl/gu, "ɭː"); // ళ్ల → geminate retroflex [ɭː] (కోళ్లు→koːɭːu)
    // First-syllable (weak) stress: mark the first vowel nucleus.
    const m = new RegExp(`[${VOWEL}]`, "u").exec(x);
    if (m && m.index !== undefined) x = x.slice(0, m.index) + "ˈ" + x.slice(m.index);
    return x.normalize("NFC");
}

const toAscii = (d: string): string =>
    [...d].map((c) => TELUGU_DIGITS[c] ?? c).join("");
/**
 * Digits → IPA. The compositor is Telugu's OWN (numbers.ts), not the shared `indicNumberWords`: Telugu
 * orders 21-99 tens-first and INFLECTS its magnitude nouns for count and for a following remainder, and
 * the shared composer expresses neither — see the numbers.ts header for the corpus/audio evidence.
 */
function number(digits: string): string {
    const n = Number(toAscii(digits));
    if (!Number.isSafeInteger(n)) return digits;
    return numberToWords(n).split(" ").map(phonemizeWord).join(" ");
}

// The foreign arm is `LATIN_RUN`, ALL of Latin plus marks — not `[A-Za-z]+`, which ended the token at a
// diacritic and left that letter to be read as an English letter name (`Cañitas` → *ka ˈɛn ˈitas*). This
// engine ROUTES a foreign word to the injected reader, so widening the class is the whole fix (#657).
const TOKEN = new RegExp(
    `([${TELUGU_WORD}]+)|(${LATIN_RUN})|([${DIGIT_CLASS}]+)|([।॥.?!,;:])`,
    "gu",
);

export type ForeignPhonemizer = (latin: string) => string;

class TeluguPhonemizer implements Phonemizer {
    constructor(private foreign?: ForeignPhonemizer) {}
    text(input: string): string {
        // TEXT NORMALIZATION runs first, before tokenization — it is pure text→text (see normalize.ts).
        return assembleClauses(normalizeTelugu(input), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            else if (m[2]) sink.emit(this.foreign ? this.foreign(m[2]) : "");
            else if (m[3]) sink.emit(number(m[3]));
            else if (m[4]) {
                const mk = CLAUSE_MARK[m[4]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Telugu phonemizer. `foreign` handles embedded Latin runs. */
export function createTelugu(foreign?: ForeignPhonemizer): Phonemizer {
    return new TeluguPhonemizer(foreign);
}

File: csharp/Vernacula.Phonemizer/Languages/Telugu/Telugu.cs
/**
 * Native Telugu (te) text phonemizer — canonical IPA.
 * Ported from src/languages/telugu/telugu.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Telugu;

public static class TeluguPhonemizer
{
    private static TeluguManifest DEF => Manifest.MANIFEST;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;
    private const string TELUGU_WORD = "ఀ-౯ౠ-ౣ";
    private static readonly IReadOnlyDictionary<string, string> TELUGU_DIGITS = new Dictionary<string, string>(StringComparer.Ordinal)
    {
        ["౦"] = "0", ["౧"] = "1", ["౨"] = "2", ["౩"] = "3", ["౪"] = "4",
        ["౫"] = "5", ["౬"] = "6", ["౭"] = "7", ["౮"] = "8", ["౯"] = "9",
    };
    private static readonly string DIGIT_CLASS = "0-9" + string.Concat(TELUGU_DIGITS.Keys);
    private const string VOWEL = "aeiouɾ"; // nucleus starts (for the geminate/stress scan; ɾ only in the ɾu vocalic-r nucleus)

    private static Func<string, string>? G2P;
    private static readonly object GATE = new();
    private static string G2p(string w)
    {
        lock (GATE) G2P ??= Abugida.MakeAbugidaG2P(DEF, PhonologyLoader.LoadSharedPhonology());
        return G2P(w);
    }

    private static readonly JsRe GEMINATE =
        JsRegex.Compile("(t͡ʃʰ|d͡ʒʱ|t͡ʃ|d͡ʒ|t͡s|d͡z|t̪ʰ|d̪ʱ|ʈʰ|ɖʱ|ɡʱ|kʰ|t̪|d̪|[kɡpbmnlʃʂsʈɖɳɭɲŋjɦʋɾr])\\1(?!͡)", "gu");
    private static readonly JsRe FINAL_ANUSVARA = JsRegex.Compile("ం$", "u");
    private static readonly JsRe ASPIRATE_AFTER_LENGTH = JsRegex.Compile("ː([ʰʱ])", "gu");
    private static readonly JsRe RETROFLEX_LL = JsRegex.Compile("ɭl", "gu");
    private static readonly JsRe FIRST_VOWEL = JsRegex.Compile($"[{VOWEL}]", "u");

    /** One Telugu word → canonical IPA. */
    public static string PhonemizeWord(string word)
    {
        var norm = JsRegex.Replace(Js.Normalize(word, System.Text.NormalizationForm.FormC), FINAL_ANUSVARA, _ => "మ్");
        var x = G2p(norm);
        x = JsRegex.Replace(x, GEMINATE, m => m.Groups[1].Value + "ː");
        x = JsRegex.Replace(x, ASPIRATE_AFTER_LENGTH, m => m.Groups[1].Value + "ː");
        x = JsRegex.Replace(x, RETROFLEX_LL, _ => "ɭː"); // ళ్ల → geminate retroflex [ɭː] (కోళ్లు→koːɭːu)
        var m2 = FIRST_VOWEL.Match(x);
        if (m2.Success) x = x[..m2.Index] + "ˈ" + x[m2.Index..];
        return x.Normalize(System.Text.NormalizationForm.FormC);
    }

    private static string ToAscii(string d) =>
        string.Concat(Js.CodePoints(d).Select(c => TELUGU_DIGITS.GetValueOrDefault(c) ?? c));

    /** Digits → IPA. */
    private static string Number(string digits)
    {
        var n = Js.Number(ToAscii(digits));
        // ⚠ Mirrors JS `Number.isSafeInteger`: above 2^53 the double has already lost the low digits, so the
        // numeral is read digit-at-a-time THROUGH THE SAME COMPOSER rather than composed as a quantity.
        if (!(double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d))
            return string.Join(" ", Js.CodePoints(ToAscii(digits))
                .SelectMany(d => TeluguNumbersComposer.NumberToWords(Js.Number(d)).Split(' '))
                .Select(PhonemizeWord));
        return string.Join(" ", TeluguNumbersComposer.NumberToWords(n).Split(' ').Select(PhonemizeWord));
    }

    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"([{TELUGU_WORD}]+)|({HostWord.LATIN_RUN})|([{DIGIT_CLASS}]+)|([।॥.?!,;:])", "gu");

    private sealed class Engine : ILanguage
    {
        private readonly Func<string, string>? _foreign;
        internal Engine(Func<string, string>? foreign) => _foreign = foreign;

        public string Text(string input) =>
            Clauses.AssembleClauses(Normalize.NormalizeTelugu(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(PhonemizeWord(m.Groups[1].Value));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0) sink.Emit(_foreign is not null ? _foreign(m.Groups[2].Value) : "");
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0) sink.Emit(Number(m.Groups[3].Value));
                else if (m.Groups[4].Success && m.Groups[4].Value.Length > 0)
                {
                    var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[4].Value);
                    if (mk is not null) sink.Pause(mk);
                }
            });
    }

    /** Build the Telugu phonemizer. `foreign` handles embedded Latin runs. */
    public static ILanguage CreateTelugu(Func<string, string>? foreign = null) => new Engine(foreign);

    internal static void RegisterSelf() =>
        Registry.Register("telugu", () => CreateTelugu(Registry.ReadAsEnglish));
}
